core/src/execution/mod.rs

```diff
@@ -1,109 +1,68 @@
-use std::collections::{HashMap, HashSet};
+use std::collections::HashMap;
+use std::sync::Arc;
 
 use alloy::consensus::BlockHeader;
+use alloy::eips::BlockId;
 use alloy::network::primitives::HeaderResponse;
-use alloy::network::{BlockResponse, ReceiptResponse};
-use alloy::primitives::{keccak256, Address, B256, U256};
-use alloy::rlp;
-use alloy::rpc::types::{BlockTransactions, Filter, FilterChanges, Log};
-use alloy_trie::root::ordered_trie_root_with_encoder;
+use alloy::network::BlockResponse;
+use alloy::primitives::{Address, B256, U256};
+use alloy::rpc::types::{
+    BlockTransactions, EIP1186AccountProofResponse, Filter, FilterChanges, Log,
+};
+use async_trait::async_trait;
 use eyre::Result;
-use futures::future::try_join_all;
-use revm::primitives::{BlobExcessGasAndPrice, KECCAK_EMPTY};
+use revm::primitives::BlobExcessGasAndPrice;
 use tracing::warn;
 
-use crate::fork_schedule::ForkSchedule;
-use crate::network_spec::NetworkSpec;
-use crate::types::BlockTag;
+use helios_common::{
+    fork_schedule::ForkSchedule,
+    network_spec::NetworkSpec,
+    types::{Account, BlockTag},
+};
 
-use self::constants::MAX_SUPPORTED_LOGS_NUMBER;
+use self::client::ExecutionInner;
 use self::errors::ExecutionError;
-use self::proof::{verify_account_proof, verify_storage_proof};
-use self::rpc::ExecutionRpc;
+use self::proof::verify_block_receipts;
+use self::spec::ExecutionSpec;
 use self::state::{FilterType, State};
-use self::types::Account;
 
+pub mod client;
 pub mod constants;
 pub mod errors;
 pub mod evm;
 pub mod proof;
 pub mod rpc;
+pub mod spec;
 pub mod state;
-pub mod types;
 
 #[derive(Clone)]
-pub struct ExecutionClient<N: NetworkSpec, R: ExecutionRpc<N>> {
-    pub rpc: R,
-    state: State<N, R>,
+pub struct ExecutionClient<N: NetworkSpec> {
+    client: Arc<dyn ExecutionInner<N>>,
+    state: State<N>,
     fork_schedule: ForkSchedule,
 }
 
-impl<N: NetworkSpec, R: ExecutionRpc<N>> ExecutionClient<N, R> {
-    pub fn new(rpc: &str, state: State<N, R>, fork_schedule: ForkSchedule) -> Result<Self> {
-        let rpc: R = ExecutionRpc::new(rpc)?;
-        Ok(ExecutionClient::<N, R> {
-            rpc,
+impl<N: NetworkSpec> ExecutionClient<N> {
+    pub fn new(
+        client: Arc<dyn ExecutionInner<N>>,
+        state: State<N>,
+        fork_schedule: ForkSchedule,
+    ) -> Result<Self> {
+        Ok(Self {
+            client,
             state,
             fork_schedule,
         })
     }
 
     pub async fn check_rpc(&self, chain_id: u64) -> Result<()> {
-        if self.rpc.chain_id().await? != chain_id {
+        if self.client.chain_id().await? != chain_id {
```

Comment:
Oh nvm I see why you have chain id in the client. This brings me back to my point on the boundary between verified and unverified data being muddled.

Reply:
Left a ToDo for this as discussed.

Reply:
Let's track it here -- https://github.com/a16z/helios/issues/544.